models/tshirt/include/Tshirt.h

```diff
@@ -0,0 +1,284 @@
+
+            // TODO: confirm these are correct
+            // Max transmissivity
+            double max_lateral_flow = params.satdk * params.multiplier * params.Ssmax;
+            double max_perc_flow = params.satdk;
+
+            // init subsurface later flow outlet
+            subsurface_outlets[lf_outlet_index] = Reservoir_Outlet(params.Klf, 1.0, Sfc, max_lateral_flow);
+            // init subsurface percolation flow outlet
+            subsurface_outlets[perc_outlet_index] = Reservoir_Outlet(params.satdk * params.slope, 1.0, Sfc, max_perc_flow);
+
+            Nonlinear_Reservoir subsurface_reservoir(0.0, params.depth, state.Ss, subsurface_outlets);
+
+            double subsurface_excess;
+            subsurface_reservoir.response_meters_per_second(subsurface_infiltration_flux, dt, subsurface_excess);
+
+            // lateral subsurface flow
+            double Qlf = subsurface_reservoir.velocity_meters_per_second_for_outlet(lf_outlet_index);
+
+            // percolation flow
+            double Qperc = subsurface_reservoir.velocity_meters_per_second_for_outlet(perc_outlet_index);
+
+            // TODO: make sure ET doesn't need to be taken out sooner
+            new_state.Ss = subsurface_reservoir.get_storage_height_meters() -
+                           calc_et(subsurface_reservoir.get_storage_height_meters(), et_params);
+
+            // initialize the Nash cascade of nonlinear reservoirs
+            std::vector<Nonlinear_Reservoir> nash_cascade;
+            // TODO: verify correctness of activation_threshold (Sfc) and max_velocity (max_lateral_flow) arg values
+            init_nash_cascade_vector(nash_cascade, params, state, Sfc, max_lateral_flow);
+
+            // cycle through lateral flow Nash cascade of nonlinear reservoirs
+            // loop essentially copied from Hymod logic, but with different variable names
+            for(unsigned long int i = 0; i < nash_cascade.size(); ++i)
+            {
+                // get response water velocity of nonlinear reservoir
+                Qlf = nash_cascade[i].response_meters_per_second(Qlf, dt, subsurface_excess);
+                // TODO: confirm this is correct
+                Qlf += subsurface_excess / dt;
+            }
+
+            // "raw" GW calculations
+            //state.Sgw += Qperc * dt;
+            //double Qgw = params.Cgw * ( exp(params.expon * state.Sgw / params.Sgwmax) - 1 );
+
+            // Given the equation:
+            //      double Qgw = params.Cgw * ( exp(params.expon * state.Sgw / params.Sgwmax) - 1 );
+            // The max value should be when Sgw == Sgwmax, or ...
+            double max_gw_velocity = params.Cgw * ( exp(params.expon) - 1 );
+            // TODO: verify activation threshold
+            Nonlinear_Reservoir groundwater_res(0, params.Sgwmax, state.Sgw, params.Cgw, 1, 0, max_gw_velocity);
+            // TODO: what needs to be done with this value?
+            double excess_gw_water;
```

Comment:
After calling groundwater_res, the excess_gw_water should be added to the fluxes.surface_runoff.

We added the excess groundwater to runoff in hymod, and I assume it would be added to runoff here too.
https://github.com/NOAA-OWP/ngen/blob/master/models/hymod/include/Hymod.h#L149

        //TODO: Review issues with dt and internal timestep
        runoff_meters_per_second += groundwater_excess_meters / dt;

The excess is in meters so be sure to divide by dt.

Reply:
Yes, essentially if there is any excess it has to go back up the stack and get added to the bucket of water that provided the original water to that call.

Reply:
@jdmattern-noaa, I don't see how that fits into things in the case of T-shirt, whereas it's more clear in Hymod, but it seems sensible.  I'll make the change, but I'll need to confirm whether that belongs in the runoff amounts before or after GIUH calculations are performed.

Reply:
@fred-ogden what to do with potentially excess water from the groundwater reservoir?

Reply:
We had a discussion about this at the end of the standup.  If a later reservoir becomes or approaches being full, that should restrict the flow from the earlier source.  Here, e.g., if the GW reservoir is full, that impose a limit on the max allowed `Qperc`.  If both the GW and the subsurface reservoirs are full, then this limits the influx into the subsurface reservoir, and the excess water gets attributed to surface runoff.

Reply:
Hmm....a stack of water!

Reply:
I am still caught up on this thought. Seems to me that things could get kind of complicated with having to go back and recalculate the Qperc to reconcile the difference with at which exact point in the timestep the GW reservoir overflows. Could we just keep Qperc as originally calculated and update the subsurface reservoir state storage height with the GW excess? If the subsurface reservoir storage height then exceeds max after the update, then that subsurface reservoir's excess could be added to the runoff.